sdks/java/io/common/src/main/java/org/apache/beam/sdk/io/common/synthetic/SyntheticBoundedInput.java

```diff
@@ -0,0 +1,452 @@
+      // generated values to not be random enough because the position values are
+      // close to each other. To make seeds fed into the Random objects unrelated,
+      // we use a hashing function to map the position to its corresponding hashcode,
+      // and use the hashcode as a seed to feed into the Random object.
+      long hashCodeOfPosition = hashFunction.hashLong(position).asLong();
+      return new Record(genKvPair(hashCodeOfPosition), nextDelay(hashCodeOfPosition));
+    }
+
+    /**
+     * Record generated by {@link #genRecord}.
+     */
+    public static class Record {
+      public final KV<byte[], byte[]> kv;
+      public final Duration sleepMsec;
+
+      public Record(KV<byte[], byte[]> kv, long sleepMsec) {
+        this.kv = kv;
+        this.sleepMsec = new Duration(sleepMsec);
+      }
+    }
+  }
+
+  /**
+   * A reader over the {@link PCollection} of {@code KV<byte[], byte[]>}
+   * from the synthetic source.
+   *
+   * <p>The random but deterministic record at position "i" in the range [A, B)
+   * is generated by using {@link SourceOptions#genRecord}.
+   * Reading each record sleeps according to the sleep time distribution
+   * in {@code SyntheticOptions}.
+   */
+  private static class SyntheticSourceReader
+      extends OffsetBasedSource.OffsetBasedReader<KV<byte[], byte[]>> {
+    private final long splitPointFrequencyRecords;
+
+    private KV<byte[], byte[]> currentKvPair;
+    private long index;
+    private boolean isAtSplitPoint;
+
+    public SyntheticSourceReader(SyntheticBoundedSource source) {
+      super(source);
+      this.currentKvPair = null;
+      this.splitPointFrequencyRecords = source.sourceOptions.splitPointFrequencyRecords;
+    }
+
+    @Override
+    public synchronized SyntheticBoundedSource getCurrentSource() {
+      return (SyntheticBoundedSource) super.getCurrentSource();
+    }
+
+    @Override
+    protected long getCurrentOffset() throws IllegalStateException {
+      return index;
+    }
+
+    @Override
+    public KV<byte[], byte[]> getCurrent() throws NoSuchElementException {
+      if (currentKvPair == null) {
+        throw new NoSuchElementException(
+            "The current element is unavailable because either the reader is "
+                + "at the beginning of the input and start() or advance() wasn't called, "
+                + "or the last start() or advance() returned false.");
+      }
+      return currentKvPair;
+    }
+
+    @Override
+    public boolean allowsDynamicSplitting() {
+      return splitPointFrequencyRecords > 0;
+    }
+
+    @Override
+    protected final boolean startImpl() throws IOException {
+      this.index = getCurrentSource().getStartOffset();
+      if (splitPointFrequencyRecords > 0) {
+        while (index % splitPointFrequencyRecords != 0) {
+          ++index;
+        }
+      }
+
+      SourceOptions options = getCurrentSource().sourceOptions;
+      SyntheticUtils.delay(options.nextInitializeDelay(this.index),
```

Comment:
Is the value of the return of this function needed for something? At first glance I thought that delays were for event time but I have the impression I misintepreted this. So maybe worth documenting on `Record`, Nexmark uses something like `TimestampedValue<Record>` to model event time. Wondering if it makes sense to add also  the implementation of  `getCurrentTimestamp()`. Which provides value even for BoundedSources.